File: tests/step11-public-hydration.py
"""Local Chrome fixtures; --live checks real logged-out reads without writes."""
from http.server import ThreadingHTTPServer, SimpleHTTPRequestHandler
from pathlib import Path
import subprocess, tempfile, threading, sys
from urllib.parse import unquote
ROOT=Path(__file__).resolve().parents[1]
LIVE='--live' in sys.argv
CLIENT=r'''
export async function getSupabaseClient(){
 const f=window.top.fixture;
 if(f.clientFail) throw Error('fixture client failure');
 return {from(table){
  const q={table,filters:[],orders:[],select(){return this;},eq(k,v){this.filters.push([k,v]);return this;},single(){this.one=true;return this;},order(k,o){this.orders.push([k,o]);return this;},async then(resolve,reject){
   try{
    window.top.reads.push({table:this.table,filters:this.filters,orders:this.orders,one:this.one});
    await new Promise(r=>setTimeout(r,80));
    if(f.networkFail)throw Error('network failure');
    if(f.fail===table)return resolve({error:{message:'private fixture error'}});
    let data=table==='homepage_content'?f.home:f.rows;
    if(Array.isArray(data)) data=[...data].filter(r=>this.filters.every(([k,v])=>r[k]===v)).sort((a,b)=>a.sort_order-b.sort_order||a.id-b.id);
    resolve({data,error:null});
   }catch(e){reject(e);}
  }};return q;
 }};
}
'''
requests=[]
finished=threading.Event()
live_result=[]
class Handler(SimpleHTTPRequestHandler):
 def __init__(self,*a,**kw):super().__init__(*a,directory=str(ROOT),**kw)
 def log_message(self,*a):pass
 def send_text(self,text,kind):
  try:
   self.send_response(200);self.send_header('Content-Type',kind);self.end_headers();self.wfile.write(text.encode('utf-8'))
  except (ConnectionAbortedError,ConnectionResetError,BrokenPipeError):pass
 def do_GET(self):
  if self.path.startswith('/test-result?'):
   live_result.append(unquote(self.path.split('?',1)[1]));finished.set();self.send_text('OK','text/plain');return
  requests.append(('GET',self.path))
  if self.path=='/test':
   self.send_text('<html><body><script>window.live='+str(LIVE).lower()+';</script><script type="module" src="/tests/step11-public-hydration.js"></script></body></html>','text/html');return
  if self.path.startswith('/repo/'):
   path=self.path.split('?',1)[0][len('/repo/'):]
   if path=='index.html':
    html=(ROOT/'index.html').read_text(encoding='utf-8')
    instrumentation='''<script>window.errors=[];window.calls=[];addEventListener('error',e=>errors.push(e.message));addEventListener('unhandledrejection',e=>errors.push(String(e.reason)));const originalFetch=window.fetch;window.fetch=(input,init)=>{const url=String(input.url||input);const method=init?.method||input.method||'GET';calls.push({url,method});if(method!=='GET'&&method!=='HEAD')return Promise.reject(Error('Non-read request blocked by test'));return originalFetch(input,init);};</script>'''
    html=html.replace('<head>','<head>'+instrumentation)
    html=html.replace('<script type="module" src="js/homepage-hydration.js"></script>','<script type="module">window.run=import("./js/homepage-hydration.js").then(m=>m.hydrationComplete);</script>')
    self.send_text(html,'text/html');return
   if path=='admin/js/supabase-client.js' and not LIVE:self.send_text(CLIENT,'text/javascript');return
   self.path='/'+path
  super().do_GET()
 def do_POST(self):requests.append(('WRITE',self.path));self.send_error(405)
 do_PUT=do_POST
 do_PATCH=do_POST
 do_DELETE=do_POST
server=ThreadingHTTPServer(('127.0.0.1',0),Handler)
threading.Thread(target=server.serve_forever,daemon=True).start()
with tempfile.TemporaryDirectory(prefix='vca-public-') as profile:
 args=['C:/Program Files/Google/Chrome/Application/chrome.exe','--headless','--disable-gpu','--no-first-run','--no-default-browser-check','--disable-background-networking','--user-data-dir='+profile]
 if LIVE:
  # Real network/image decoding needs wall-clock time, not virtual-time dumping.
  process=subprocess.Popen(args+[f'http://127.0.0.1:{server.server_port}/test'],stdout=subprocess.DEVNULL,stderr=subprocess.DEVNULL)
  try:
   finished.wait(40)
   result='<pre id="test-results">'+(live_result[0] if live_result else 'BLOCKED: Live browser checks did not finish within 40 seconds.')
  finally:
   process.terminate();process.wait(timeout=10)
 else:
  run=subprocess.run(args+['--dump-dom','--virtual-time-budget=30000',f'http://127.0.0.1:{server.server_port}/test'],capture_output=True,text=True,encoding='utf-8',timeout=45)
  start=run.stdout.find('<pre id="test-results">')
  result=run.stdout[start:run.stdout.find('</pre>',start)]
 print(result);print('Non-GET local requests:',[r for r in requests if r[0]!='GET'])
 server.shutdown()
 if 'test-results">PASS' not in result or any(r[0]!='GET' for r in requests):
  print('Chrome did not finish the checks successfully.');raise SystemExit(1)

File: tests/step11-public-hydration.js
import {defaultContent} from '/admin/js/editor-defaults.js';
const results=[];
const check=(name,ok)=>{if(!ok)throw Error(name);results.push(name);};
const wait=ms=>new Promise(r=>setTimeout(r,ms));
const textFields=['hero_eyebrow','hero_title','hero_description','about_label','about_title','about_description','footer_text'];
const imgFields={'site-logo':'site_logo_url','hero-image':'hero_image_url','about-image':'about_image_url'};
window.reads=[];
const fresh=()=>({home:{...structuredClone(defaultContent)},rows:[{...defaultContent.announcements[0],id:1}]});
let frame;
async function open(f){
 frame?.remove();window.fixture=f;
 frame=document.createElement('iframe');frame.src='/repo/index.html';document.body.append(frame);
 await new Promise(r=>frame.onload=r);
 for(let i=0;!frame.contentWindow.run&&i<200;i++)await wait(10);
 return frame.contentWindow;
}
const node=name=>frame.contentDocument.querySelector(`[data-cms="${name}"]`);
const style=()=>frame.contentDocument.documentElement.style;
const unchanged=()=>node('hero-title').textContent.trim()==='VCA Philippines'&&node('announcements').children.length===1&&node('announcement-title').textContent.includes('Welcome to the VCA')&&node('hero-button').textContent.trim()==='Learn More'&&!node('hero-button').hidden&&!style().cssText&&node('hero-image').getAttribute('src')==='assets/hero-default.webp';
try{
 if(window.live){
  const w=await open(null);check('Live logged-out hydration succeeds',await w.run);
  const {loadPublicHomepage}=await w.eval('import("/repo/js/public-data.js")');const data=await loadPublicHomepage();
  check('Fresh browser has no auth session',!Object.keys(w.localStorage).some(k=>k.includes('auth-token')));
  check('All saved text appears',textFields.every(f=>node(f.replaceAll('_','-')).textContent===data.homepage[f]));
  for(const [target,field] of Object.entries(imgFields)){
   const img=node(target);img.loading='eager';await img.decode();
   check('Live '+target+' loaded',img.src===new URL(data.homepage[field],frame.contentDocument.baseURI).href&&img.naturalWidth>0);
  }
  check('Saved button applied',data.homepage.hero_button_text?node('hero-button').textContent===data.homepage.hero_button_text&&node('hero-button').href===new URL(data.homepage.hero_button_link,frame.contentDocument.baseURI).href:node('hero-button').hidden);
  check('All announcements in saved order',[...node('announcements').querySelectorAll('h2')].map(n=>n.textContent).join('|')===data.announcements.map(r=>r.announcement_title).join('|'));
  check('Saved theme colors applied',[['theme_primary_color','--vca-blue'],['theme_secondary_color','--vca-sky'],['theme_background_color','--background'],['theme_text_color','--text']].every(([f,p])=>style().getPropertyValue(p)===data.homepage[f]));
  const stack='system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", sans-serif';
  const named={'poppins':'Poppins','montserrat':'Montserrat','inter':'Inter','roboto':'Roboto','open-sans':'Open Sans','lato':'Lato','merriweather':'Merriweather'};
  check('Saved font stacks applied',[['theme_heading_font','--heading-font'],['theme_body_font','--body-font']].every(([f,p])=>style().getPropertyValue(p)===(data.homepage[f]==='system-default'?stack:`"${named[data.homepage[f]]}", ${stack}`)));
  check('Live network only reads with no auth actions',w.calls.length>=2&&w.calls.every(c=>c.method==='GET'&&!c.url.includes('/auth/')&&!c.url.includes('/storage/')));
  check('Live no uncaught errors',!w.errors.length);
 }else{
  let f=fresh();textFields.forEach(k=>f.home[k]='<b>'+k+'</b>');
  f.home.hero_button_text='<b>Go</b>';f.home.hero_button_link='pages/welcome.html';
  f.home.theme_primary_color='#123456';f.home.theme_secondary_color='#234567';f.home.theme_background_color='#345678';f.home.theme_text_color='#456789';
  f.rows=[{...f.rows[0],id:8,sort_order:2,announcement_title:'Third'},{...f.rows[0],id:7,sort_order:1,announcement_title:'Second'},{...f.rows[0],id:2,sort_order:1,announcement_title:'<b>First</b>'}];
  let w=await open(f);check('Static content visible before reads resolve',unchanged());check('Hydration succeeds',await w.run);
  check('All text targets hydrate literally',textFields.every(k=>node(k.replaceAll('_','-')).textContent===f.home[k]));
  check('No HTML injection',!frame.contentDocument.querySelector('b'));
  check('Button text and repository-relative URL',node('hero-button').textContent==='<b>Go</b>'&&node('hero-button').href.endsWith('/repo/pages/welcome.html'));
  for(const [target,field] of Object.entries(imgFields))check(target+' relative path and preserved alt',node(target).src===new URL(f.home[field],frame.contentDocument.baseURI).href&&!!node(target).alt);
  check('Four color variables applied',['--vca-blue','--vca-sky','--background','--text'].every((p,i)=>style().getPropertyValue(p)===['#123456','#234567','#345678','#456789'][i]));
  check('Background applies to main surfaces',[frame.contentDocument.body,frame.contentDocument.querySelector('.site-header'),node('announcements').firstChild].every(n=>w.getComputedStyle(n).backgroundColor==='rgb(52, 86, 120)'));
  check('System font maps exactly',style().getPropertyValue('--body-font')==='system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", sans-serif');
  check('All announcements sorted by order then ID',[...node('announcements').querySelectorAll('h2')].map(n=>n.textContent).join('|')==='<b>First</b>|Second|Third');
  check('SELECT filters and ascending ordering',reads.some(r=>r.table==='homepage_content'&&r.one&&r.filters[0][0]==='id'&&r.filters[0][1]===1)&&reads.some(r=>r.table==='homepage_announcements'&&r.filters[0][0]==='homepage_id'&&r.filters[0][1]===1&&JSON.stringify(r.orders)==='[["sort_order",{"ascending":true}],["id",{"ascending":true}]]'));
  const {prepareHydration}=await w.eval('import("/repo/js/homepage-hydration.js")');
  for(const [token,family] of Object.entries({poppins:'Poppins',montserrat:'Montserrat',inter:'Inter',roboto:'Roboto','open-sans':'Open Sans',lato:'Lato',merriweather:'Merriweather'})){
   f.home.theme_heading_font=token;f.home.theme_body_font=token;prepareHydration({homepage:f.home,announcements:[]})();check(token+' font allowlist',style().getPropertyValue('--heading-font').startsWith('"'+family+'", system-ui')&&style().getPropertyValue('--body-font')===style().getPropertyValue('--heading-font'));
  }
  for(const link of ['#about','https://example.org/path','http://example.org/','mailto:hello@example.org','tel:+1234']){f.home.hero_button_link=link;prepareHydration({homepage:f.home,announcements:[]})();check('Button protocol '+link,node('hero-button').href===new URL(link,frame.contentDocument.baseURI).href);}
  f=fresh();f.home.hero_button_text='';f.home.hero_button_link='';f.rows=[];w=await open(f);await w.run;
  check('Empty button hidden accessibly',node('hero-button').hidden&&w.getComputedStyle(node('hero-button')).display==='none');check('Successful zero rows clears sample',node('announcements').children.length===0);
  f=fresh();f.home.hero_button_link='javascript:alert(1)';f.home.site_logo_url='blob:bad';f.home.hero_image_url='file:///C:/bad';f.home.about_image_url='data:image/png,x';f.home.theme_primary_color='red; background:url(x)';f.home.theme_body_font='toString';
  w=await open(f);await w.run;check('Dangerous button retains authored fallback',node('hero-button').getAttribute('href')==='#about');check('Unsafe images retain authored sources',Object.entries(imgFields).every(([t,k])=>node(t).getAttribute('src')===defaultContent[k]));check('Invalid theme values retain defaults',!style().getPropertyValue('--vca-blue')&&!style().getPropertyValue('--body-font'));
  f=fresh();f.home.hero_image_url='https://example.org/saved.webp';w=await open(f);await w.run;check('HTTPS persisted image assigned',node('hero-image').src==='https://example.org/saved.webp');
  for(const [label,change] of [['Homepage failure',f=>f.fail='homepage_content'],['Announcement failure',f=>f.fail='homepage_announcements'],['Missing singleton',f=>f.home=null],['Client/CDN failure',f=>f.clientFail=true],['Network failure',f=>f.networkFail=true],['Malformed homepage',f=>f.home.hero_title={}],['Malformed announcements',f=>f.rows={}],['Malformed row',f=>f.rows=[null]]]){
   f=fresh();change(f);w=await open(f);check(label+' preserves full fallback',!(await w.run)&&unchanged());check(label+' handled without uncaught errors',!w.errors.length);
  }
  f=fresh();w=await open(f);node('footer-text').removeAttribute('data-cms');check('Missing DOM target prevents partial application',!(await w.run)&&unchanged());
  check('Fixture network has no writes',w.calls.every(c=>c.method==='GET'));
 }
 frame?.remove();const out=document.createElement('pre');out.id='test-results';out.textContent='PASS\n'+results.join('\n');document.body.replaceChildren(out);
}catch(e){const out=document.createElement('pre');out.id='test-results';out.textContent='FAIL\n'+e.stack+'\nPassed: '+results.join(', ');document.body.replaceChildren(out);}

if(window.live)await fetch("/test-result?"+encodeURIComponent(document.getElementById("test-results").textContent));
